### PR TITLE
Add UUID Sync Tests

### PR DESCRIPTION
**Pull request by dianaafanador3**
This PR adds 
* Sync tests for objectId/UUID/String/int primary keys objects
* Sync tests for objectId/UUID/String/int partition values types .
* Support for querying using UUID as primaryKey
 `realm.object(ofType: SwiftPrimaryUUIDObject.self, forPrimaryKey: UUID(uuidString: "4ee1fa48-8b23-11eb-8dcd-0242ac130003")!)`
* Additional object test for Swift/Int/UUID as primary keys
* Additional test for indexable UUID

Updated baas commit to use latest version to 31/03/2021
realm-Core branch: v11

### CODE REVIEW

File: Realm/ObjectServerTests/RLMSyncTestCase.h [hunk 1]
```diff
@@ -40,6 +39,7 @@ NS_ASSUME_NONNULL_BEGIN
 @property NSString *firstName;
 @property NSString *lastName;
 
+- (instancetype) initWithPrimaryKey:(RLMObjectId *)primaryKey age:(NSInteger)strCol firstName:(NSString *)intCol lastName:(NSString *)intCol;
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RLMSyncTestCase.h:42]: No space between the type and the name in obj-c methods.

File: Realm/ObjectServerTests/RLMSyncTestCase.mm [hunk 2]
```diff
@@ -200,9 +296,14 @@ - (RLMUser *)anonymousUser {
 }
 
 - (RLMCredentials *)basicCredentialsWithName:(NSString *)name register:(BOOL)shouldRegister {
+    return [self basicCredentialsWithName:name register:shouldRegister app:nil];
+}
+
+- (RLMCredentials *)basicCredentialsWithName:(NSString *)name register:(BOOL)shouldRegister app:(nullable RLMApp*) app {
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RLMSyncTestCase.mm:302]: Space before \* in obj-c

File: Realm/ObjectServerTests/RLMSyncTestCase.mm [hunk 3]
```diff
@@ -200,9 +296,14 @@ - (RLMUser *)anonymousUser {
...
     if (shouldRegister) {
         XCTestExpectation *expectation = [self expectationWithDescription:@""];
-        [self.app.emailPasswordAuth registerUserWithEmail:name password:@"password" completion:^(NSError *error) {
+        RLMApp *currentApp = app != nil ? app : self.app;
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RLMSyncTestCase.mm:305]: `app ?: self.app`

File: Realm/ObjectServerTests/RLMSyncTestCase.mm [hunk 4]
```diff
@@ -559,6 +639,55 @@ - (NSTask *)childTask {
     return [self childTaskWithAppIds:_appId ? @[_appId] : @[]];
 }
 
+- (RLMApp *)createAppForPartition:(id<RLMBSON>)partition {
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RLMSyncTestCase.mm:642]: This being `createAppForPartition:` confused me a bit, as I expected it to be doing something which actually required an app-per-partition, while it appears to only need an app per partition key type. I think making it `createAppForPartitionType:` instead would be a bit clearer.

File: Realm/ObjectServerTests/RLMSyncTestCase.mm [hunk 5]
```diff
@@ -559,6 +639,55 @@ - (NSTask *)childTask {
...
+    syncManager.userAgent = self.name;
+
+    return _app;
+}
+
+- (NSString *)partitionBsonType:(RLMBSONType)type {
+    switch(type){
+        case RLMBSONTypeString:
+            return @"string";
+            break;
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RLMSyncTestCase.mm:674]: breaks here are redundant

File: Realm/ObjectServerTests/RealmServer.swift [hunk 6]
```diff
@@ -518,7 +518,7 @@ public class RealmServer: NSObject {
     }
 
     /// Create a new server app
-    @objc public func createApp() throws -> AppId {
+    @objc public func createApp(forBSONType bsonType: String) throws -> AppId {
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RealmServer.swift:521]: Swift naming usually doesn't use `for` here.

File: Realm/ObjectServerTests/RealmServer.swift [hunk 7]
```diff
@@ -684,25 +686,110 @@ public class RealmServer: NSObject {
             "relationships": [:]
         ]
 
-        let userDataRule: [String: Any] = [
+        let uUIDPrimaryKeyObjectRule: [String: Any] = [
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RealmServer.swift:689]: `uuidPrimaryKeyObjectRule`

Initialisms always have the same capitalization for the entire thing. In obj-c they're always all uppercase unless they're at the start of a local variable name. In Swift they're uppercase unless they're at the start of any non-type name.

**Comment by dianaafanador3** [comment]
@tgoyne comments addressed, PTAL

File: Realm/ObjectServerTests/RLMObjectServerPartitionTests.mm [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+@interface RLMObjectServerPartitionTests : RLMSyncTestCase
+@end
+
+@implementation RLMObjectServerPartitionTests
+
+- (void)roundTripForPartitionValue:(id<RLMBSON>)value  {
+    NSString *appId;
+    if (self.isParent) {
+        NSError *error;
+        appId = [RealmServer.shared createAppForBSONType: [self partitionBsonType:value]  error:&error];
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RLMObjectServerPartitionTests.mm:46]: No space after colon for obj-c message sends.

File: Realm/ObjectServerTests/RLMObjectServerPartitionTests.mm [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+        if (error) {
+            XCTFail(@"Could not create app for partition value %@d", value);
```
Comment (tgoyne) [inline Realm/ObjectServerTests/RLMObjectServerPartitionTests.mm:49]: You'll want to also return after this so that it doesn't try to run the actual test.

**Review by tgoyne — APPROVED** [review]

**Pull request merged**